lib/rdoc/parser/prism_ruby.rb

```diff
@@ -0,0 +1,975 @@
+    case container
+    when RDoc::ClassModule then
+      @top_level.add_to_classes_or_modules container
+    end
+
+    container.record_location @top_level
+  end
+
+  # Scans this Ruby file for Ruby constructs
+
+  def scan
+    @tokens = RDoc::Parser::RipperStateLex.parse(@content)
+    @lines = @content.lines
+    result = Prism.parse(@content)
+    @program_node = result.value
+    @line_nodes = {}
+    prepare_line_nodes(@program_node)
+    prepare_comments(result.comments)
+    return if @top_level.done_documenting
+
+    @first_non_meta_comment = nil
+    if (_line_no, start_line, rdoc_comment = @unprocessed_comments.first)
+      @first_non_meta_comment = rdoc_comment if start_line < @program_node.location.start_line
+    end
+
+    @program_node.accept(RDocVisitor.new(self, @top_level, @store))
+    process_comments_until(@lines.size + 1)
+  end
+
+  def should_document?(code_object) # :nodoc:
+    return true unless @track_visibility
+    return false if code_object.parent&.document_children == false
+    code_object.document_self
+  end
+
+  # Assign AST node to a line.
+  # This is used to show meta-method source code in the documentation.
+
+  def prepare_line_nodes(node) # :nodoc:
+    case node
+    when Prism::CallNode, Prism::DefNode
+      @line_nodes[node.location.start_line] ||= node
+    end
+    node.compact_child_nodes.each do |child|
+      prepare_line_nodes(child)
+    end
+  end
+
+  # Prepares comments for processing. Comments are grouped into consecutive.
+  # Consecutive comment is linked to the next non-blank line.
+  #
+  # Example:
+  #   01| class A # modifier comment 1
+  #   02|   def foo; end # modifier comment 2
+  #   03|
+  #   04|   # consecutive comment 1 start_line: 4
+  #   05|   # consecutive comment 1 linked to line: 7
+  #   06|
+  #   07|   # consecutive comment 2 start_line: 7
+  #   08|   # consecutive comment 2 linked to line: 10
+  #   09|
+  #   10|   def bar; end # consecutive comment 2 linked to this line
+  #   11| end
+
+  def prepare_comments(comments)
+    current = []
+    consecutive_comments = [current]
+    @modifier_comments = {}
+    comments.each do |comment|
+      if comment.is_a? Prism::EmbDocComment
+        consecutive_comments << [comment] << (current = [])
+      elsif comment.location.start_line_slice.match?(/\S/)
+        @modifier_comments[comment.location.start_line] = RDoc::Comment.new(comment.slice, @top_level, :ruby)
+      elsif current.empty? || current.last.location.end_line + 1 == comment.location.start_line
+        current << comment
+      else
+        consecutive_comments << (current = [comment])
+      end
+    end
+
+    # Example: line_no = 5, start_line = 2, comment_text = "# comment_start_line\n# comment\n"
+    # 1| class A
+    # 2|   # comment_start_line
+    # 3|   # comment
+    # 4|
+    # 5|   def f; end # comment linked to this line
+    # 6| end
+    @unprocessed_comments = consecutive_comments.reject(&:empty?).map do |comments|
```

Comment:
We could avoid allocating multiple arrays by changing this to:

```suggestion
    consecutive_comments.reject!(&:empty?)
    @unprocessed_comments = consecutive_comments.map! do |comments|
```

I normally consider this micro-optimization. But given RDoc could have long consecutive comments, I feel this may help in some cases.